applications/tari_base_node/src/parser.rs

```diff
@@ -20,21 +20,105 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+use super::LOG_TARGET;
+use crate::builder::Interfaces;
+use log::*;
+use rustyline::{
+    completion::Completer,
+    error::ReadlineError,
+    hint::{Hinter, HistoryHinter},
+    line_buffer::LineBuffer,
+    Context,
+};
+use rustyline_derive::{Helper, Highlighter, Validator};
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
 use tokio::runtime;
 
 /// This is used to parse commands from the user and execute them
+#[derive(Helper, Validator, Highlighter)]
 pub struct Parser {
     executor: runtime::Handle,
+    interfaces: Interfaces,
+    shutdown_flag: Arc<AtomicBool>,
+    commands: Vec<&'static str>,
+    hinter: HistoryHinter,
+}
+
+// This will go through all instructions and look for potential matches
+impl Completer for Parser {
+    type Candidate = String;
+
+    fn complete(&self, line: &str, pos: usize, _ctx: &Context<'_>) -> Result<(usize, Vec<String>), ReadlineError> {
+        let mut completions: Vec<String> = Vec::new();
+        for command in &self.commands {
+            if command.starts_with(line) {
+                completions.push(command.to_string());
+            }
+        }
+
+        Ok((pos, completions))
+    }
+
+    fn update(&self, line: &mut LineBuffer, start: usize, elected: &str) {
+        line.update(elected, start);
+    }
+}
+
+// This allows us to make hints based on historic inputs
+impl Hinter for Parser {
+    fn hint(&self, line: &str, pos: usize, ctx: &rustyline::Context<'_>) -> Option<String> {
+        self.hinter.hint(line, pos, ctx)
+    }
 }
 
 impl Parser {
     /// creates a new parser struct
-    pub fn new(executor: runtime::Handle) -> Self {
-        Parser { executor }
+    pub fn new(executor: runtime::Handle, interfaces: Interfaces, shutdown_flag: Arc<AtomicBool>) -> Self {
+        Parser {
+            executor,
+            interfaces,
+            shutdown_flag,
+            commands: vec!["help", "get wallet balance", "send tari", "get basenode meta", "quit"],
```

Comment:
Don't put spaces in commands
```suggestion
            commands: vec!["help", "get_balance", "send_tari", "get_chain_metadata", "quit", "exit"],
```